code/datums/ai/ai_controller.dm

```diff
@@ -0,0 +1,753 @@
+	 * you're adding a datum reference to this! Use the setters, this is
+	 * important for reference handing.
+	 */
+	var/list/blackboard = list()
+
+	/// Bitfield of traits for this AI to handle extra behavior.
+	var/ai_traits = NONE
+	/// Current actions planned to be performed by the AI in the upcoming plan.
+	var/list/planned_behaviors
+	/// Current actions being performed by the AI.
+	var/list/current_behaviors
+	/// Current actions and their respective last time ran as an assoc list.
+	var/list/behavior_cooldowns = list()
+	/// Current status of AI (OFF/ON)
+	var/ai_status
+	/// Current movement target of the AI, generally set by decision making.
+	var/atom/current_movement_target
+	/// Identifier for what last touched our movement target, so it can be cleared conditionally
+	var/movement_target_source
+	/// Stored arguments for behaviors given during their initial creation
+	var/list/behavior_args = list()
+	/// Tracks recent pathing attempts, if we fail too many in a row we fail our current plans.
+	var/consecutive_pathing_attempts
+	/// Can the AI remain in control if there is a client?
+	var/continue_processing_when_client = FALSE
+	/// Distance to give up on target.
+	var/max_target_distance = 14
+	/// Cooldown for new plans, to prevent AI from going nuts if it can't think of new plans and looping on end
+	COOLDOWN_DECLARE(failed_planning_cooldown)
+	/// All subtrees this AI has available. Will run them in order, so make sure
+	/// they're in the order you want them to run. On initialization of this
+	/// type, it will start as a typepath(s) and get converted to references of
+	/// ai_subtrees found in SSai_controllers when init_subtrees() is called
+	var/list/planning_subtrees
+
+	/// The idle behavior this AI performs when it has no actions.
+	var/datum/idle_behavior/idle_behavior = null
+
+	// Movement related things here
+	/// Reference to the movement datum we use. Is a type on initialize but becomes a ref afterwards.
+	var/datum/ai_movement/ai_movement
+	/// Delay between movements. This is on the controller so we can keep the movement datum singleton
+	var/movement_delay = 0.2 SECONDS
+
+	// TODO: Move the variables below into the blackboard at some point.
+
+	/// AI paused time
+	var/paused_until = 0
+	/// Can this AI idle?
+	var/can_idle = TRUE
+	/// What distance should we be checking for interesting things when considering idling/deidling? Defaults to AI_DEFAULT_INTERESTING_DIST
+	var/interesting_dist = AI_DEFAULT_INTERESTING_DIST
+
+/datum/ai_controller/New(atom/new_pawn)
+	change_ai_movement_type(ai_movement)
+	init_subtrees()
+
+	if(idle_behavior)
+		idle_behavior = new idle_behavior()
+
+	if(!isnull(new_pawn)) // unit tests need the ai_controller to exist in isolation due to list schenanigans i hate it here
+		possess_pawn(new_pawn)
+
+/datum/ai_controller/Destroy(force)
+	set_ai_status(AI_STATUS_OFF)
+	unpossess_pawn(FALSE)
+	set_movement_target(type, null)
```

Comment:
Are all of the lists holding datums cleared through these procs?

Reply:
The subtrees, behaviors and targeting strategies are singletons stored on SSai_controllers and SSai_behaviors, so we don't want to qdel the list contents, but I'll `LAZYCLEARLIST` them here just in case.